src/jit/smallhash.h

```diff
@@ -0,0 +1,591 @@
+        {
+            unsigned precedingIndexInChain = index;
+
+            index = (index + offset) & mask;
+            bucket = &m_buckets[index];
+
+            assert(bucket->m_isFull);
+            if (bucket->m_hash == hash && TKeyInfo::Equals(bucket->m_key, key))
+            {
+                *precedingIndex = precedingIndexInChain;
+                *bucketIndex = index;
+                return true;
+            }
+        }
+
+        return false;
+    }
+
+    //------------------------------------------------------------------------
+    // HashTableBase::Resize: allocates a new bucket array twice the size of
+    //                        the current array and copies the key-value pairs
+    //                        from the current bucket array into the new array.
+    void Resize()
+    {
+        Bucket* currentBuckets = m_buckets;
+
+        unsigned newNumBuckets = m_numBuckets == 0 ? InitialNumBuckets : m_numBuckets * 2;
+        size_t allocSize = sizeof(Bucket) * newNumBuckets;
+        assert((sizeof(Bucket) * m_numBuckets) < allocSize);
+
+        auto* newBuckets = reinterpret_cast<Bucket*>(m_compiler->compGetMem(allocSize));
+        memset(newBuckets, 0, allocSize);
+
+        for (unsigned currentIndex = 0; currentIndex < m_numBuckets; currentIndex++)
+        {
+            Bucket* currentBucket = &currentBuckets[currentIndex];
+            if (!currentBucket->m_isFull)
+            {
+                continue;
+            }
+
+            bool inserted = Insert(newBuckets, newNumBuckets, currentBucket->m_hash, currentBucket->m_key, currentBucket->m_value);
+            (assert(inserted), (void)inserted);
+        }
+
+        m_numBuckets = newNumBuckets;
+        m_buckets = newBuckets;
+    }
+
+protected:
+    HashTableBase(Compiler* compiler, Bucket* buckets, unsigned numBuckets)
+        : m_compiler(compiler)
+        , m_buckets(buckets)
+        , m_numBuckets(numBuckets)
+        , m_numFullBuckets(0)
+    {
+        assert(compiler != nullptr);
+
+        if (numBuckets > 0)
+        {
+            assert((numBuckets & (numBuckets - 1)) == 0); // Size must be a power of 2
+            assert(m_buckets != nullptr);
+
+            memset(m_buckets, 0, sizeof(Bucket) * numBuckets);
+        }
+    }
+
+public:
+    class Iterator;
+
+    class KeyValuePair final
+    {
+        friend class HashTableBase<TKey, TValue, TKeyInfo>::Iterator;
+
+        Bucket* m_bucket;
+
+        KeyValuePair(Bucket* bucket)
+            : m_bucket(bucket)
+        {
+            assert(m_bucket != nullptr);
+        }
+
+    public:
+        KeyValuePair()
+            : m_bucket(nullptr)
+        {
+        }
+
+        inline TKey& Key()
+        {
+            return m_bucket->m_key;
+        }
+
+        inline TValue& Value()
+        {
+            return m_bucket->m_value;
+        }
+    };
+
+    class Iterator final
```

Comment:
Perhaps. They're currently only used for dumping the contents of a table in debug builds.